src/EditorFeatures/CSharp/CodeCleanup/CSharpCodeCleanupService.cs

```diff
@@ -0,0 +1,181 @@
+                }
+            };
+
+            return dictionary.ToImmutableDictionary();
+        }
+
+        public async Task<Document> CleanupDocument(Document document, CancellationToken cancellationToken)
+        {
+            var docOptions = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+
+            document = await ApplyCodeFixes(document, docOptions, cancellationToken).ConfigureAwait(false);
+            // do the remove usings after code fix, as code fix might remove some code which can results in unused usings.
+            document = await RemoveSortUsings(document, docOptions, cancellationToken).ConfigureAwait(false);
+
+            return document;
+        }
+
+        private async Task<Document> RemoveSortUsings(Document document, DocumentOptionSet docOptions, CancellationToken cancellationToken)
+        {
+            // remove usings
+            if (docOptions.GetOption(FeatureOnOffOptions.RemoveUnusedUsings))
+            {
+                var removeUsingsService = document.GetLanguageService<IRemoveUnnecessaryImportsService>();
+                if (removeUsingsService != null)
+                {
+                    document = await removeUsingsService.RemoveUnnecessaryImportsAsync(document, cancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            // sort usings
+            if (docOptions.GetOption(FeatureOnOffOptions.SortUsings))
+            {
+                document = await OrganizeImportsService.OrganizeImportsAsync(document, cancellationToken).ConfigureAwait(false);
+            }
+
+            return document;
+        }
+
+        private async Task<Document> ApplyCodeFixes(Document document, DocumentOptionSet docOptions, CancellationToken cancellationToken)
+        {
+            var fixAllService = document.Project.Solution.Workspace.Services.GetService<IFixAllGetFixesService>();
+
+            var dummy = new ProgressTracker();
+            foreach (var diagnosticId in GetEnabledDiagnosticIds(docOptions))
+            {
+                var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                var textSpan = new TextSpan(0, syntaxTree.Length);
+
+                var fixCollection = await _codeFixService.GetFixesAsync(document, textSpan, diagnosticId, cancellationToken).ConfigureAwait(false);
+                if (fixCollection == null)
+                {
+                    continue;
+                }
+
+                var fixAll = fixCollection.FixAllState;
+                var solution = await fixAllService.GetFixAllChangedSolutionAsync(
+                    fixAll.CreateFixAllContext(dummy, cancellationToken)).ConfigureAwait(false);
+
+                document = solution.GetDocument(document.Id);
+            }
+
+            return document;
+        }
+
+        private List<string> GetEnabledDiagnosticIds(DocumentOptionSet docOptions)
```

Comment:
will this be called a lot?  Consider an approach that uses a pooled list.

Reply:
It gets called once each time Click+K, D is pressed, which I doubt will be very often
We will optimize it if it starts getting called often